### DISCUSSION TITLE
Can I create a profile for an E3D ToolChanger with four printheads?

### DISCUSSION THREAD

**Discussion by terryrobison** [OrcaSlicer/OrcaSlicer#1376] [General]
Can I create a profile for an E3D ToolChanger with four printheads?  The print heads (tools) have different nozzle sizes.  I use superslicer for this printer but would like to use OrcaSlicer.  

I can't see how I can add additional tools, and swap between them as required.  I have looked at start g-code, but I can only address a single tool that way.  I am sure I am approaching this the wrong way.  Any pointers?

Regards

**Comment by TexZeTech** [comment]
@terryrobison 

I'm assuming you are running Klipper, if that is the case you should be able to handle it all with a macro similar to the one here.

https://github.com/ld3300/Klipper_Configs/blob/master/configs/extruders.cfg

I do not have a printer with that setup.... yet... so good luck.

**Comment by terryrobison** [comment]
Thank you for your response TexZeTech.  I am running a DUET 2 with a recent
version of RepRatFirmware.  RepRat does support macros (everything is
gcode).  I couldn't figure out how to select Tool0 or Tool3 (I have 4
printheads) in OrcaSlicer.  I can do this in SuperSlicer easily.  Each Tool
can be a different-sized nozzle etc.  Perhaps I missed where to define
multiple tool heads in OrcaSlicer.  I have looked many times, but not sure
where this might be.   There is a an IDEX Duet board RepRat in the list of
printers.  I had a look at that profile, but it looks like it only handles
one tool at a time, not both at the same time.

Regards

On Sat, 5 Aug 2023 at 08:46, TexZeTech ***@***.***> wrote:

> @terryrobison <https://github.com/terryrobison>
>
> I'm assuming you are running Klipper, if that is the case you should be
> able to handle it all with a macro similar to the one here.
>
> https://github.com/ld3300/Klipper_Configs/blob/master/configs/extruders.cfg
>
> I do not have a printer with that setup.... yet... so good luck.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SoftFever/OrcaSlicer/discussions/1376#discussioncomment-6642925>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIV76NCRSDFEJUCOOWK4YDLXTV3UHANCNFSM6AAAAAAZKX2WLY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Terry

"You can log off / But you can never leave."
 --------------------------------------------------------------
 -  Terry Robison
 -  Senior Systems Analyst
 -  Blueprint I.T. Services Pty. Ltd.
 -  ***@***.***         _\|/_
 -                                                   (o o)
 -------------------------------------oOO-(_)-OOo----

**Reply by Festivejelly** [reply]
Bit late to the party here but couldnt you do this in the filament before and after gcode?

So like have 4 filaments and filament one would select T0, filament 2 select T1 etc etc

Might need to do some sort of post processing to remove unneeded stuff at the beginning of prints etc.

The problem is that it wont let you do things like choose which extruder for outer walls etc. But for most stuff you should be able to get by,

**Comment by broncosis** [comment]
@terryrobison   actually reprap firmware on a duet 2 support macros very well I use them lots on my railcore

You make a gcode file and name it what you want the macro to be. put it on your SD card

**Reply by psychotik2k3** [reply]
The thing that none of you answered is can he use multiple tools, some of them having different nozzle size... that changes a lot of things.
i wanted to try orca slicer and couldn't find how i could setup my 4 tools (different extruder, speeds, retraction etc) but it seems orca slicer is not ready for that.

**Reply by TexZeTech** [reply]
You are correct that its not a click n' go but as I stated above a person could make it happen via macros. 

eg; (this macro doesn't work and is just for general concept)

```
if tool#5
{
nozzle_diameter[0.6]
}
if tool#1
{
nozzle_diameter[0.4]
}
```

So ya its not polished for multi tool as I and others would like compared to say tools that are geared towards but it can be done via macros.

**Reply by psychotik2k3** [reply]
I'm not sure to follow you. Where would be that macro? Orca slicer or klipper/rrf config?

**Reply by TexZeTech** [reply]
https://github.com/TypQxQ/Klipper_ToolChanger

https://github.com/rosscon/klipper-macros/

https://github.com/supermerill/SuperSlicer/discussions/4157

Honestly you could make it all orcaslicer side. It is dependent on your setup.

**Reply by psychotik2k3** [reply]
Ok so it's macro on the printer side. So this does not work because the path for a 0.4 nozzle won't be the same as a 0.6 or 0.2 nozzle. This has to be set on the slicer side.

This will work only if your using sale nozzle size, same kind of materials (because flow is not the same for you than pla) cooling etc.

**Reply by TexZeTech** [reply]
Take a look at the variables in the slicer (there is an editor if you haven't found it). I think its doable but you have to be the judge of that yourself.

**Reply by psychotik2k3** [reply]
Which variables are there (I abandoned orca since it can't handle my 3 extruders) ?
Because the problem is the slicing algorithm that won't take the extruder size in account to compute the path.

**Reply by TexZeTech** [reply]
Looking though the issues section, people have already requested Multi tool support. If  you want a plug & play option at this point in time its not Orcaslicer. 

I did state I think its doable with macros and I stand by that however I haven't built a macro for myself as of yet.

I also don't know why you couldn't download and check the variables yourself but here ya go.

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/94c1d37d-9f2c-497c-b573-88779cf4cd75" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/f9648a24-3d53-4187-8ac2-45af141f8738" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/0495b282-e98b-4fb9-928a-3407a7403896" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/af038827-7592-492f-bcac-6f0ec90e0b00" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/2476ace0-75c3-4364-b4cd-2e241ec01c64" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/124300cf-5a12-42e3-a85d-47c4ada4706b" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/6035044b-555b-43b7-baa3-3947833a3fc1" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/642929d8-da11-4635-8432-2e092c49eb75" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/bfb742ac-f630-4a10-a1b3-1ef15e9203cf" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/282c40f8-2c46-4fd4-b4e5-d927a3bfb7ea" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/97bd0974-36bc-46ff-85ec-e309b37fd75d" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/5791c9ea-50af-471f-b0f0-f74f63ccea61" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/e838960a-2da8-4d9a-89e9-8e7addd2b51d" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/38e0d6f1-8445-44ea-9e27-2ba031058589" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/b7230232-c7cc-4bda-8603-692695c674c4" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/ca907491-8049-4db5-9350-535386d87191" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/12093bd8-7458-4eee-9bfc-2736d901d32e" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/99c17a08-4713-4f4a-af8f-b50f2137e464" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/a67f5a5f-6e53-4b1e-811c-133e8b0d29e9" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/3a3dc7e1-bb60-4282-b160-1c261c44f844" width=10% height=10%>

<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/eac16eee-e994-4694-aecb-f6ebaf65a832" width=10% height=10%>
